cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -123,12 +132,43 @@ protected void runOneIteration() throws Exception {
     if (threadPoolExecutor != null) {
       for (EventReader<StartProgramEvent> reader : readers) {
         threadPoolExecutor.execute(() -> {
-          processEvents(reader);
+          if (runRecordMonitorService.isRunning()) {
+            // Only attempt to process event if there is no max or the current count is less than max
+            if (hasNominalCapacity()) {
+              processEvents(reader);
+            }
+          } else {
+            LOG.warn("RunRecordMonitorService not running, in state: {}", runRecordMonitorService.state());
```

Comment:
Change to "RunRecordMonitorService not yet running, currently in state: {} . Status will be checked again in next attempt." . This will give more clarity that this most likely will resolve soon.